db/migrate/20200813031851_add_unique_index_to_profiles.rb

```diff
@@ -0,0 +1,8 @@
+class AddUniqueIndexToProfiles < ActiveRecord::Migration[6.0]
+  disable_ddl_transaction!
+
+  def change
+    remove_index :profiles, column: :user_id, algorithm: :concurrently
+    add_index :profiles, :user_id, unique: true, algorithm: :concurrently
```

Comment:
Yes! :)

Reply:
Will it cause any problems that these are both running concurrently or does postgres know to remove one and then add the next? Or is it fine if both end up running at the same time? Just curious I am not super educated on Postgres internals :)

Reply:
Great question!

> Regular index builds permit other regular index builds on the same table to occur simultaneously, but only one concurrent index build can occur on a table at a time. In either case, schema modification of the table is not allowed while the index is being built. Another difference is that a regular CREATE INDEX command can be performed within a transaction block, but CREATE INDEX CONCURRENTLY cannot.

from https://www.postgresql.org/docs/11/sql-createindex.html#SQL-CREATEINDEX-CONCURRENTLY

as I interpret it, the second index is queued and has to wait.

As the index is done outside of a transaction and is entered a catalog before it is actually built, I think we're safe here:

> In a concurrent index build, the index is actually entered into the system catalogs in one transaction, then two table scans occur in two more transactions. Before each table scan, the index build must wait for existing transactions that have modified the table to terminate. After the second scan, the index build must wait for any transactions that have a snapshot (see Chapter 13) predating the second scan to terminate. Then finally the index can be marked ready for use, and the CREATE INDEX command terminates. Even then, however, the index may not be immediately usable for queries: in the worst case, it cannot be used as long as transactions exist that predate the start of the index build.

If we weren't, we could not even do this in different files. We'd have to wait for one index to be finished before submitting the second one (can't wait for PostgreSQL 12 so we can build a mini admin panel to monitor concurrent index building and its status: https://www.depesz.com/2019/04/18/waiting-for-postgresql-12-report-progress-of-create-index-operations/)

Reply:
@rhymes Thanks for looking this up.

Additionally, I'd like to add that `CONCURRENTLY` doesn't refer to indexing operations (as quoted you can't actually have 2 different `CREATE INDEX CONCURRENTLY` on the same table at the same time) but update operations on the indexed table, avoiding potential locks. The `DROP` and `CREATE` will most likely happen in sequence, `CONCURRENTLY` just ensures that we don't lock the table, which is a bit of a moot point anyway, since the table will be empty before this PR gets deployed.